app/presenters/api/service_plans_presenter.rb

```diff
@@ -1,8 +1,8 @@
 # frozen_string_literal: true
 
 class Api::ServicePlansPresenter < PlansBasePresenter
-  def initialize(service:, user:, params: {})
-    super(collection: service.provider.service_plans, user: user, params: params)
+  def initialize(service:, collection:, user:, params: {})
+    super(collection: collection, user: user, params: params)
```

Comment:
I don't think we need to pass collection, just adjust the scope:
```suggestion
  def initialize(service:, user:, params: {})
    super(collection: service.service_plans, user: user, params: params)
```

As a side note, `#collection` does pretty much the same as `Api::PlansBaseController#find_service` so it could be cleaned up .

Reply:
I guess the idea here is that the presenter is not responsible for figuring out what elements of the collection to show, it's the controller's responsibility.

Like, if there was some restriction on what plans the current user can see, the filtering would be made by the controller, and not the presenter.

Reply:
I tried to do
```
  def collection
    service.service_plans
  end
```

in `Api::ServicePlansController`, but there is an issue with the order of [these two hooks](https://github.com/3scale/porta/blob/scope-service-plans-by-product/app/controllers/api/plans_base_controller.rb#L10-L11).

In order for some controller methods to work like this, `find_service` and `find_plan` would need to be swapped. But some others are broken after this swapping, specifically updating the plan (as its URL does not include the `service_id`, so `service_id` is [derived from the service plan](https://github.com/3scale/porta/blob/scope-service-plans-by-product/app/controllers/api/plans_base_controller.rb#L56))

It would be easy if all URLs would be prefixed by service ID, but it's not the case.

So, in my opinion, this refactoring is not worth it, as it may break some existing functionality. Or well, we can do it in another PR, as it's not necessary for fixing this issue.